rfd/9999-webauthn-support.md

```diff
@@ -0,0 +1,674 @@
+```
+
+Authentication (also referred simply as Login) follows a similar "challenge ->
+sign -> verify" protocol. In simple terms, the client requests a 
+[CredentialAssertion](https://pkg.go.dev/github.com/duo-labs/webauthn/protocol#CredentialAssertion)
+from the server, signs it by calling
+[navigartor.credentials.get()](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer/get),
+and replies with a
+[CredentialAssertionResponse](https://pkg.go.dev/github.com/duo-labs/webauthn/protocol#CredentialAssertionResponse).
+Once again, assuming all is well, login is complete.
+
+See https://webauthn.io/ and https://webauthn.guide/ for starter references.
+
+### U2F/WebAuthn compatibility
+
+WebAuthn is backwards compatible with U2F, but it's important to examine what
+this sentence actually means.
+
+U2F, as the current register/login protocol implemented by Teleport, is not
+forward compatible with WebAuthn - WebAuthn messages differ from its
+predecessor, including various options that didn't exist before. Browser support
+is implemented via the
+[navigator.credentials](https://developer.mozilla.org/en-US/docs/Web/API/CredentialsContainer)
+object, which is WebAuthn-exclusive and
+[distinct from the U2F APIs in use](https://github.com/google/u2f-ref-code#reference-code-for-u2f-specifications).
+
+WebAuthn is backwards compatible with U2F in the sense that FIDO/CTAP1 devices,
+ie U2F-capable devices, are still supported. The main difference in
+implementations lies in the definition of the App ID / Relying Party ID (ie,
+"https://example.com:3080" vs "example.com"). The "appid" extension is the
+solution for this problem and seems to be well-supported for browsers
+(__TODO(codingllama):__ Verify for Firefox and Safari).
+
+Unfortunately, the appid extension is not without its issues.
+[As the standard itself says in item 2](https://www.w3.org/TR/webauthn/#sctn-appid-extension),
+_"When verifying the assertion, expect that the rpIdHash MAY be the hash of the
+AppID instead of the RP ID."_ This advice
+[wasn't followed by github.com/duo-labs/webauthn](https://github.com/duo-labs/webauthn/blob/9f1b88ef44cc0e4f5ddf511ed12a3aa468f972d7/protocol/assertion.go#L117),
+but in the short-term may be circumvented by creating a secondary
+[WebAuthn object](https://pkg.go.dev/github.com/duo-labs/webauthn/webauthn#WebAuthn)
+with RPID = AppID, used exclusively to validate legacy requests (identified by
+the presence of the "incorrect" rpIdHash).
+
+### UX and configuration
+
+WebAuthn support is designed to be as seamless as the U2F experience. Users
+enabling MFA for the first time may register their devices via `tsh mfa add` and
+use `tsh login` as usual. Users migrating from U2F to WebAuthn retain their
+previously-registered devices and keep their workflow unchanged.
+
+Migration from U2F to WebAuthn is achieved simply by changing the second_factor
+option from "u2f" to "webauthn". The WebAuthn backend sends challenges for all
```

Comment:
@codingllama if webauthn is a superset of `u2f`, why not make update automatic, making u2f mean the same as webauthn and removing the u2f from the documentation in the future?

Reply:
The main issue I see are Teleport version rollbacks and their corner-cases. If you upgrade to a version where "u2f" means "webauthn" and later rollback, the devices registered in-between may not be true U2F devices (they may disappear, not work, etc, depending on how we go about this). I am concerned this might break user expectation if they never touched their configs during the update (vs someone who actually changed from "u2f" to "webauthn").

Let's say we do this, then the question is what are the acceptable "breakages" in case of a rollback.

* FIDO2 registered devices would "disappear" from storage (as they are saved as WebAuthnDevices)
* U2F devices registered after the upgrade would either disappear (if saved as WebAuthnDevice) or would require special system handling to keep working (ie, saving as both WebAuthnDevice and U2FDevice)

Newer devices (eg, yubikey 5+) are all likely to disappear, even if they do support U2F, as I assume they would favor the FIDO2 protocol over U2F/CTAP1.

One could also argue that this is due to happen anyway when u2f is removed/deprecated, in which case we might do it now. Devices disappearing on rollbacks would also be the current behavior for "second_factor: on".

WDYT?

Reply:
@codingllama sounds good, for the transition period let's say for 8.0 we can keep `webauthn` and `u2f` modes separate, but will replace all mentions in the documentation with webauthn going forward. Then in the next major verison we should make them aliases. If you agree, let's note this in the RFD